### PR TITLE
Allow applications to customize how roles are decoded from raw identity

### LINKED ISSUE #3123: Allow applications to customize how roles are decoded from raw identity
**Description**
During the authentication, the `elytron-security` extension is decoding roles expecting that they are always associated with a `groups` attribute after the identity was successfully authorized.

In some cases, it might be useful for applications to define how roles are decoded from the attributes associated with an identity instead of assuming that roles should always be fetched from a `groups` attribute.

This flexibility should help other extensions to avoid implementing unnecessary code if they need to decode roles differently than what is current provided.

A good example is token-based authentication, where the roles can be mapped from any claim within the token or a combination of them.

**Implementation ideas**

Define a `DefaultRoleDecoder` bean that is capable of delegating to an application-specific implementation of `RoleDecoder`, if provided.

### PR DESCRIPTION
**Pull request by pedroigor**
Fixes #3123

### CODE REVIEW

**Review by gsmet — CHANGES_REQUESTED** [review]
There is a formatting error.

Please run `./mvnw process-sources` and amend your commit.

**Comment by gsmet** [comment]
@pedroigor could you run the formatter and amend your commit, please?

Thanks!

**Review by gsmet — CHANGES_REQUESTED** [review]
Added 2 comments, could you take care of them?

FYI next release is on Wednesday at 12pm.

File: extensions/elytron-security/runtime/src/main/java/io/quarkus/elytron/security/runtime/DefaultRoleDecoder.java [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+    private Instance<RoleDecoder> instances;
+
+    @Override
+    public Roles decodeRoles(AuthorizationIdentity authorizationIdentity) {
+        RoleDecoder delegate = getDelegate();
+        
+        if (delegate == null) {
+            Attributes.Entry groups = authorizationIdentity.getAttributes().get(DEFAULT_ATTRIBUTE_NAME);
+            Set<String> roles = new HashSet<>(groups);
+            return new Roles() {
```
[the remark this answers is not in the document]
Comment (pedroigor) [reply extensions/elytron-security/runtime/src/main/java/io/quarkus/elytron/security/runtime/DefaultRoleDecoder.java]: I  realized that `Roles` provides a `fromSet` method that serves to the same purpose. Thanks for catching this.

File: extensions/elytron-security/runtime/src/main/java/io/quarkus/elytron/security/runtime/DefaultRoleDecoder.java [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+public class DefaultRoleDecoder implements RoleDecoder {
+
+    private static final String DEFAULT_ATTRIBUTE_NAME = "groups";
+    
+    @Inject
+    private Instance<RoleDecoder> instances;
+
+    @Override
+    public Roles decodeRoles(AuthorizationIdentity authorizationIdentity) {
+        RoleDecoder delegate = getDelegate();
```
[the remark this answers is not in the document]
Comment (pedroigor) [reply extensions/elytron-security/runtime/src/main/java/io/quarkus/elytron/security/runtime/DefaultRoleDecoder.java]: Yeah, but if we use the constructors we are assuming that app-defined instances are application-scoped and I'm not sure if that is always the case.

Or are you thinking about something else that I may be missing?

[the remark this answers is not in the document]
Comment (gsmet) [reply extensions/elytron-security/runtime/src/main/java/io/quarkus/elytron/security/runtime/DefaultRoleDecoder.java]: Hmmm. Right.

[the remark this answers is not in the document]
Comment (pedroigor) [reply extensions/elytron-security/runtime/src/main/java/io/quarkus/elytron/security/runtime/DefaultRoleDecoder.java]: If you think that we should worry about performance now, maybe we could check whether or not the first produced instance is application-scoped and then cache it?

Or maybe all this logic to use application provided instances into instances created during the build steps could be managed by Quarkus somehow? For instance, one drawback with the approach I'm using is that we may want to allow applications to provide instances for `RoleMapper` or `PermissionMapper`. For each of them we would need a similar instance as `DefaultRoleDecoder` just to use `Instance` to obtain the delegate.

**Review by gsmet — APPROVED** [review]

**Comment by gsmet** [comment]
I force pushed a rebase due to conflicts with the recent s/template/recorder/ changes.

Let's wait to see what CI has to say.

**Pull request merged**